ui/src/main/java/edu/wpi/grip/ui/python/PythonEditorController.java

```diff
@@ -0,0 +1,242 @@
+    codeArea.setParagraphGraphicFactory(LineNumberFactory.get(codeArea));
+    codeArea.richChanges()
+        .filter(ch -> !ch.getInserted().equals(ch.getRemoved()))
+        .subscribe(change -> codeArea.setStyleSpans(0, computeHighlighting(codeArea.getText())));
+    codeArea.replaceText(0, 0, PythonScriptFile.TEMPLATE);
+    codeArea.setStyle("-fx-font-family: 'Consolas'; -fx-font-size: 10pt;");
+    codeArea.getStylesheets()
+        .add(getClass().getResource("python-keywords.css").toExternalForm());
+    root.setCenter(codeArea);
+  }
+
+  /**
+   * Computes highlighting for python code.
+   *
+   * @param text the text to highlight
+   */
+  private static StyleSpans<Collection<String>> computeHighlighting(String text) {
+    Matcher matcher = PATTERN.matcher(text);
+    int lastKwEnd = 0;
+    StyleSpansBuilder<Collection<String>> spansBuilder = new StyleSpansBuilder<>();
+    while (matcher.find()) {
+      String styleClass =
+          matcher.group("KEYWORD") != null ? "keyword" :
+              matcher.group("PAREN") != null ? "paren" :
+                  matcher.group("BRACE") != null ? "brace" :
+                      matcher.group("BRACKET") != null ? "bracket" :
+                          matcher.group("STRING") != null ? "string" :
+                              matcher.group("COMMENT") != null ? "comment" :
+                                  null; /* never happens */
+      assert styleClass != null;
+      spansBuilder.add(Collections.emptyList(), matcher.start() - lastKwEnd);
+      spansBuilder.add(Collections.singleton(styleClass), matcher.end() - matcher.start());
+      lastKwEnd = matcher.end();
+    }
+    spansBuilder.add(Collections.emptyList(), text.length() - lastKwEnd);
+    return spansBuilder.create();
+  }
+
+  /**
+   * Gets the window displaying the editor.
+   */
+  private Window getWindow() {
+    if (window == null) {
+      window = root.getScene().getWindow();
+    }
+    return window;
+  }
+
+  /**
+   * Gets the root {@code BorderPane} for the editor.
+   */
+  public @Nonnull BorderPane getRoot() {
+    return root;
+  }
+
+  /**
+   * Gets the script text in the editor.
+   */
+  public @Nullable String getScript() {
+    if (scriptFileName() == null) {
+      return null;
+    } else {
+      return codeArea.getText();
+    }
+  }
+
+  /**
+   * Creates the name of the file to save the script to based on the operation name.
+   */
+  private @Nullable String scriptFileName() {
+    final Pattern namePattern = Pattern.compile("name *= *\"(.*)\" *");
+    String code = codeArea.getText();
+    String[] lines = code.split("\n");
+    for (String line : lines) {
+      Matcher m = namePattern.matcher(line);
+      if (m.matches()) {
+        return m.group(1).replaceAll("[ \\t]+", "_").toLowerCase(Locale.ENGLISH) + ".py";
+      }
+    }
+    return null;
+  }
+
+  @FXML
+  private void save() {
+    if (scriptFile == null) {
+      String fileName = scriptFileName();
+      if (fileName == null) {
```

Comment:
Might just want to check if `PythonScriptFile.create(scriptFileName())` throws an exception, that would also check for errors in the script